build_tools/circle/build_doc.sh

```diff
@@ -58,14 +58,62 @@ get_build_type() {
         return
     fi
     changed_examples=$(echo "$filenames" | grep -E "^examples/(.*/)*plot_")
+
+    # The following is used to extract the list of filenames of example python
+    # files that sphinx-gallery needs to run to generate png files used as
+    # figures or images in the .rst files  from the documentation.
+    # If the contributor changes a .rst file in a PR we need to run all
+    # the examples mentioned in that file to get sphinx build the
+    # documentation without generating spurious warnings related to missing
+    # png files.
+
+    if [[ -n "$filenames" ]]
+    then
+        # get rst files
+        rst_files="$(echo "$filenames" | grep -E "rst$")"
+
+        # get lines with figure or images
+        img_fig_lines="$(echo "$rst_files" | xargs grep -shE "(figure|image)::")"
+
+        # get only auto_examples
+        auto_example_files="$(echo "$img_fig_lines" | grep auto_examples | awk -F "/" '{print $NF}')"
+
+        # remove "sphx_glr_" from path and accept replace _\d\d\d.png with .py
+        scripts_names="$(echo "$auto_example_files" | sed 's/sphx_glr_//' | sed -e 's/_[[:digit:]][[:digit:]][[:digit:]].png/.py/')"
+
+        # get unique values
+        examples_in_rst="$(echo "$scripts_names" | uniq )"
```

Comment:
What about:

```bash
if [[ -n "$examples_in_rst" ]]
then
	changed_examples="$changed_examples|$examples_in_rst"
fi
```

And leave the rest unchanged. My goal is to reduce the number of `if` statements after this point.

Reply:
Then if `$changed_examples` is empty your regular expression for `EXAMPLES_PATTERN` will begin with `|`: this generates an error in every checker of regexp I've tested (maybe you know why?) and always crashes my checks in [cmarmo#5](https://github.com/cmarmo/scikit-learn/pull/5)

Reply:
I agree. Is our Makefile tolerant to extra `|` when either `changed_examples` or `examples_in_rst` are empty?

Reply:
Would this verbose way work?

```bash
if [[ -n "$examples_in_rst" ]]
then
	if [[ -n "$changed_examples" ]]
	then
		changed_examples="$changed_examples|$examples_in_rst"
	else
		changed_examples="$examples_in_rst"
	fi
fi
```

Reply:
Yes, I go for it.
Next time you write the code and I'll do the review? I feel like the entire process will be way more efficient... ;)